REGRESSION(282648@main): A tweet with 12000+ emojis takes Safari forever to display

shallawa opened the pull request:
#### 50da6f7257606b97cdc913028889c6d274583812
<pre>
REGRESSION(282648@main): A tweet with 12000+ emojis takes Safari forever to display
<a href="https://bugs.webkit.org/show_bug.cgi?id=285576">https://bugs.webkit.org/show_bug.cgi?id=285576</a>
<a href="https://rdar.apple.com/142638658">rdar://142638658</a>

Reviewed by Ryosuke Niwa.

Inserting a replaced Element (image, canvas, input, etc) should not affect the
directionality of the parent since it should be treated as a neutral character.

If the inserted elememt is an HTMLImageElement and its style has text content,
we can still skip this element for the directionality computation since its text
should not affect or interact with the rest of the text in the page.

* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::childrenChanged):
* Source/WebCore/dom/Element.h:
(WebCore::Element::isReplaced const):
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::isReplaced const):
* Source/WebCore/html/HTMLCanvasElement.h:
* Source/WebCore/html/HTMLFrameElement.h:
* Source/WebCore/html/HTMLIFrameElement.h:
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::isReplaced const):
* Source/WebCore/html/HTMLImageElement.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::isReplaced const):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/HTMLPlugInElement.cpp:
(WebCore::HTMLPlugInElement::isReplaced const):
* Source/WebCore/html/HTMLPlugInElement.h:
* Source/WebCore/html/HTMLVideoElement.h:
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::m_intrinsicSize):
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::shouldApplyLayoutContainment):
(WebCore::shouldApplySizeContainment):
(WebCore::shouldApplyInlineSizeContainment):
(WebCore::shouldApplyPaintContainment):
* Source/WebCore/svg/SVGSVGElement.cpp:
(WebCore::SVGSVGElement::isReplaced const):
* Source/WebCore/svg/SVGSVGElement.h:

Canonical link: <a href="https://commits.webkit.org/301333@main">https://commits.webkit.org/301333@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/26ed34821d56f14ffdbfd4d8427cb7468c0bf368

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125623 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45286 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36036 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132484 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77507 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e1d87b7b-3251-4893-bb08-3a8a526aa7d2) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127494 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45973 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53848 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95701 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/63828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/65eba779-cc58-430f-a166-83fd11d95f12) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128571 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36750 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112339 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76196 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/69b228b5-ccac-4492-b24f-a5b4f17059e8) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35651 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30520 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75957 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106529 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30738 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135155 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52420 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40185 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104171 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52865 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108550 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103900 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26463 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49260 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27565 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/49639 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52315 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58114 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51663 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55016 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53359 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3566,8 +3566,10 @@ void Element::childrenChanged(const ChildChange& change)
     if (document().isDirAttributeDirty()) [[unlikely]] {
-        if (selfOrPrecedingNodesAffectDirAuto())
-            updateEffectiveTextDirection();
+        if (change.type != ChildChange::Type::ElementInserted || !(is<HTMLImageElement>(change.siblingChanged) || change.siblingChanged->isReplaced())) {
```
How about other image-like elements like object, embed, iframe, etc...?

shallawa replied:
This should be covered by `isReplaced()`. All these elements return `true` in their `isReplaced()`.

shallawa committed:
REGRESSION(282648@main): A tweet with 12000+ emojis takes Safari forever to display
https://bugs.webkit.org/show_bug.cgi?id=285576
rdar://142638658

Reviewed by Ryosuke Niwa.

Inserting a replaced Element (image, canvas, input, etc) should not affect the
directionality of the parent since it should be treated as a neutral character.

If the inserted elememt is an HTMLImageElement and its style has text content,
we can still skip this element for the directionality computation since its text
should not affect or interact with the rest of the text in the page.

* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::childrenChanged):
* Source/WebCore/dom/Element.h:
(WebCore::Element::isReplaced const):
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::isReplaced const):
* Source/WebCore/html/HTMLCanvasElement.h:
* Source/WebCore/html/HTMLFrameElement.h:
* Source/WebCore/html/HTMLIFrameElement.h:
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::isReplaced const):
* Source/WebCore/html/HTMLImageElement.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::isReplaced const):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/HTMLPlugInElement.cpp:
(WebCore::HTMLPlugInElement::isReplaced const):
* Source/WebCore/html/HTMLPlugInElement.h:
* Source/WebCore/html/HTMLVideoElement.h:
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::m_intrinsicSize):
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::shouldApplyLayoutContainment):
(WebCore::shouldApplySizeContainment):
(WebCore::shouldApplyInlineSizeContainment):
(WebCore::shouldApplyPaintContainment):
* Source/WebCore/svg/SVGSVGElement.cpp:
(WebCore::SVGSVGElement::isReplaced const):
* Source/WebCore/svg/SVGSVGElement.h:

Canonical link: https://commits.webkit.org/301333@main
